ksqldb-engine/src/main/java/io/confluent/ksql/util/BinPackedStreamsMetadata.java

```diff
@@ -0,0 +1,199 @@
+                       final Map<String, Object> streamsProperties,
+                       final BinPackedPersistentQueryMetadataImpl binPackedPersistentQueryMetadata,
+                       final QueryId queryId) {
+    this.errorClassifier = errorClassifier;
+    this.streamsProperties =
+        ImmutableMap.copyOf(
+            Objects.requireNonNull(streamsProperties, "streamsProperties"));
+    this.metadata.put(queryId, binPackedPersistentQueryMetadata);
+    this.listeners.put(queryId, binPackedPersistentQueryMetadata.getListener());
+  }
+
+  protected StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse uncaughtHandler(
+      final Throwable e
+                                                                                         ) {
+    QueryError.Type errorType = QueryError.Type.UNKNOWN;
+    try {
+      errorType = errorClassifier.classify(e);
+    } catch (final Exception classificationException) {
+      LOG.error("Error classifying unhandled exception", classificationException);
+    } finally {
+      // If error classification throws then we consider the error to be an UNKNOWN error.
+      // We notify listeners and add the error to the errors queue in the finally block to ensure
+      // all listeners and consumers of the error queue (e.g. the API) can see the error. Similarly,
+      // log in finally block to make sure that if there's ever an error in the classification
+      // we still get this in our logs.
+      final QueryError queryError =
+          new QueryError(
+              System.currentTimeMillis(),
+              Throwables.getStackTraceAsString(e),
+              errorType
+          );
+      for (Map.Entry<QueryId, QueryMetadata.Listener> listener: listeners.entrySet()) {
+        listener.getValue().onError(metadata.get(listener.getKey()), queryError);
+      }
+      queryErrors.add(queryError);
+      LOG.error(
+          "Unhandled exception caught in streams thread {}. ({})",
+          Thread.currentThread().getName(),
+          errorType,
+          e
+      );
+    }
+    return StreamsUncaughtExceptionHandler.StreamThreadExceptionResponse.REPLACE_THREAD;
+  }
+
+  public KafkaStreams getKafkaStreams() {
+    return kafkaStreams;
+  }
+
+  public KafkaStreams.State state() {
+    return kafkaStreams.state();
+  }
+
+  public Collection<StreamsMetadata> allMetadata() {
+    return kafkaStreams.allMetadata();
+  }
+
+  public Set<StreamsTaskMetadata> getTaskMetadata() {
+    return kafkaStreams.localThreadsMetadata()
+                       .stream()
+                       .flatMap(t -> t.activeTasks().stream())
+                       .map(StreamsTaskMetadata::fromStreamsTaskMetadata)
+                       .collect(Collectors.toSet());
+  }
+
+  public void addToplogy(final Topology topology) {
+    queryMap.put(topology.getName(), topology);
+    kafkaStreams.addTopology(topology);
+  }
+
+  public void stop(final QueryId queryId) {
+    queryMap.remove(queryId.toString());
```

Comment:
I feel that interchangingly use `queryId.toString()` and `topology.getName()` may confused future readers that do not know we always treat them the same. Maybe just stick with one and even better just have one map. See my other comment above.

Reply:
They should both be the same but I see what you are saying, if I expand the map I might be able to only use the queryID string as the key and that should simplify things